### ISSUE TITLE
[FG:InPlacePodVerticalScaling] Implement version skew handling for in-place pod resize

### ISSUE THREAD

**Issue by vinaykul** [kubernetes/kubernetes#117767]
### What would you like to be added?

This is a tracking bug to implement two changes to better handle version skew as documented in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#version-skew-strategy
1. kubelet: When feature-gate is disabled, if kubelet sees a Proposed resize, it rejects the resize as Infeasible.
2. scheduler: If PodStatus.Resize field is not empty, scheduler uses max(ResourcesAllocated, Requests) even if the feature-gate is disabled.

### Why is this needed?

This is needed to gracefully handle component version skews that can occur during upgrades.

**Comment by Sajiyah-Salat** [comment]
Hello @infdahai infdahai, are you working on this?

**Comment by Sajiyah-Salat** [comment]
According to description of pr. The changes are documented but implementation is still remaining, right?

**Comment by infdahai** [comment]
@Sajiyah-Salat yeah, I just quit this.

**Comment by phs-sakshi** [comment]
can I work in this issue?

**Comment by infdahai** [comment]
> can I work on this issue?

you can just use `/assign` command

**Comment by samyakjain10** [comment]
Hey everyone. Just saw this issue assigned to multiple people. Was curious to know if someone is working on it or is it up for grabs?

**Comment by samyakjain10** [comment]
Hey everyone. Just saw this issue assigned to multiple people. Was curious to know if someone is working on it or is it up for grabs?

**Comment by Bhanuteja01** [comment]
@vinaykul  can you explain how to tackle with this issue

**Comment by pbialon** [comment]
I don't understand the description (sorry for that, I'm a new contributor).
> kubelet: When feature-gate is disabled, if kubelet sees a Proposed resize, it rejects the resize as Infeasible.

Is it a desired state description or a current state? If a current state, then what should be a desired action here?

**Comment by pbialon** [comment]
I've fixed (I guess) the first part in `kubelet` - @vinaykul please take a look when you're free. 
If it is not what was expected here, please guide me 🙇

**Comment by pbialon** [comment]
@vinaykul hi, could you take a look? :)

**Comment by pbialon** [comment]
@Shrihariii dunno if it makes sense for you to work on it - there is already an open pull request and discussion in it is going on.

**Comment by userAdityaa** [comment]
'''if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
		// Handle pod resize here instead of doing it in HandlePodUpdates because
		// this conveniently retries any Deferred resize requests
		// TODO(vinaykul,InPlacePodVerticalScaling): Investigate doing this in HandlePodUpdates + periodic SyncLoop scan

		//     See: https://github.com/kubernetes/kubernetes/pull/102884#discussion_r663160060
		if kl.podWorkers.CouldHaveRunningContainers(pod.UID) && !kubetypes.IsStaticPod(pod) {
			pod = kl.handlePodResourcesResize(pod)
		}

	} else {
		if kl.podWorkers.CouldHaveRunningContainers(pod.UID) && !kubetypes.IsStaticPod(pod) {
			pod.Status.Conditions = append(pod.Status.Conditions, v1.PodCondition{
				Type:   v1.PodConditionType("ResizeInfeasible"),
				Status: v1.ConditionTrue,
				Reason: "InPlacePodVerticalScaling feature gate is disabled",
			})
		}
	}'''

If the feature gate is disabled simply we can set the pod status conditions to be infeasible ?

**Comment by tallclair** [comment]
This is blocked by the decision of how to handle version skew. See https://github.com/kubernetes/kubernetes/issues/126616.

TL;DR - at this point, any version skew solution that depends on Kubelet will push out the release another ~3 releases, which is unacceptable.

**Comment by tallclair** [comment]
/remove-good-first-issue
/remove-help

This isn't a good first issue, since the logic is still up for debate. See https://github.com/kubernetes/enhancements/pull/4704 for the discussion.

I don't think anyone is actively working on this, please add yourself back and reach out to me if you are.

/unassign @amomsi
/unassign @pbialon
/unassign @itiwarishubham
/unassign @archits581
/unassign @devilcrises
/unassign @wlp1153468871
/unassign @Shrihariii
/unassign @EtherSol
/unassign @Bhanuteja01
/unassign @MeenuyD

**Comment by sreeram-venkitesh** [comment]
I'm interested in taking this up. https://github.com/kubernetes/kubernetes/issues/125394 seems to be related to how we handle the version skew policy, how do we go about that?

**Comment by tallclair** [comment]
The version skew strategy is based on the resources field, not the resize policy, so I think https://github.com/kubernetes/kubernetes/issues/125394 is not tied to this, unless I'm missing something?

**Comment by tallclair** [comment]
Oh, you were referring to the side note about version skew? I think that's out of date, and can be ignored.
